allennlp/training/trainer.py

```diff
@@ -1111,4 +611,74 @@ def from_params(cls,  # type: ignore
                    log_batch_size_period=log_batch_size_period)
 
 
-Trainer.register("default")(Trainer)
+class TrainerPieces(NamedTuple):
+    """
+    These are things that need to get instantiated from_params
```

Comment:
It'd be nicer if this said _why_ this exists this way.  This loads the datasets, constructs a vocabulary, then passes that to the `Model`, making `Trainer.from_params` need less custom logic, and so that if some other `Trainer` has similar needs, it can re-use this.